DS-56 [가게] Controller 클래스 설계 및 정의

y00njinuk opened the pull request:
## Goals
- 가게 컨트롤러 클래스를 설계 및 정의한다.

## Implements
- [x] 응답 데이터로 활용되는 DTO 구조체에 누락된 필드 추가
- [x] 가게 컨트롤러 클래스 정의 및 구현
- [x] 예외 메세지 및 핸들러 스펙 추가
- [x] Json 직렬화/역직렬화용 Bean 추가
- [x] 가게 컨트롤러 클래스 테스트 추가   

## Related Issue 
- https://dangol-sonnim.atlassian.net/browse/DS-56

## Test
- "새로운 가게 정보 생성을 요청하면 정상적으로 생성이 된다."
- "특정 가게 아이디 값을 통해 가게를 검색할 수 있다."
- "특정 가게 정보를 업데이트하여 정보를 수정할 수 있다."
- "존재하지 않은 가게 정보로 업데이트를 요청하면 Bad Request를 반환받는다."

y00njinuk committed:
feat: add missing fields of StoreResponseDTO

[issue-key] DS-56
- 응답 데이터로 활용되는 DTO 구조체에 누락된 필드 추가

y00njinuk committed:
feat: add configuration for json mapper

[issue-key] DS-56
- JSON Serde 용도로 사용할 Bean 객체 추가

y00njinuk committed:
feat: add ErrorCodeMessage "REQUEST_NOT_INVALID"

[issue-key] DS-56
- JSON 형식의 데이터가 적절하지 않은 경우의 에러 메세지 스펙 추가

y00njinuk committed:
feat: add controller of Store

[issue-key] DS-56
- 가게와 관련된 신규 컨트롤러 추가

y00njinuk committed:
Merge branch 'main' into feature/DS-56-add-store-controller

# Conflicts:
#	src/main/java/com/dangol/dangolsonnimbackend/errors/enumeration/ErrorCodeMessage.java

y00njinuk committed:
rafactor: add to handler RuntimeException

[issue-key] DS-56
- RuntimeException 핸들러 추가

y00njinuk committed:
test: check to invalid request "/api/v1/store/update"

[issue-key] DS-56
- RuntimeException 핸들러 추가

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/dto/StoreResponseDTO.java:
```diff
@@ -21,6 +21,8 @@ public class StoreResponseDTO {
 
     private String comments;
 
+    private String sido;
```
Response 데이터로 제공할 DTO 객체에 일부 누락된 필드가 있어서 추가합니다.

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/config/JsonMapperConfig.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class JsonMapperConfig {
```
Json 직렬화/역직렬화 용도로 사용할 Bean을 추가하여 사용할 수 있도록 Configuration 어노테이션을 할당한 클래스를 추가하였습니다.

rnjstmdals6 replied:
ResponseEntity가 dto 객체를 json 형식으로 직렬화하고
@RequestBody 가 dto 객체를 json 형식으로 역직렬화하지 않나요?

y00njinuk replied:
일반적인 경우에는 말씀해주신게 맞는데 Optional 필드에 대하여 직렬화/역직렬화를 할 수 있도록 Jdk8Module을 등록하는 스펙을 추가하였습니다.

해당 스펙을 추가함으로써 다음을 기대할 수 있습니다.
- 존재하지 않는 키/값에 대하여 null 값 생략
- Optional 필드에 대한 직렬화/역직렬화 기능 추가

rnjstmdals6 replied:
Optional 필드가 어떻게 사용될 지 궁금하네요.
아직 DTO에 적용은 안하신거죠?

y00njinuk replied:
StoreSignupRequestDTO에 적용이 되어 있는 상태입니다.

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/errors/enumeration/ErrorCodeMessage.java:
```diff
@@ -14,6 +14,7 @@ public enum ErrorCodeMessage {
     PASSWORD_NOT_MATCH("패스워드가 일치하지 않습니다."),
     BOSS_NOT_FOUND("존재하지 않는 사장님입니다."),
     STORE_NOT_FOUND("존재하지 않는 가게입니다."),
+    REQUEST_NOT_INVALID("요청 데이터의 JSON 형식이 적절하지 않습니다."),
```
음.. 이 부분은 JSON 데이터 형식이 맞지 않거나 문제가 될 때 발생할 수 있는 Exception에 대한 스펙을 추가하였습니다.
테스트에 추가하고 싶은데 아직은 활용을 못해봤네요.

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/errors/RestApiExceptionHandler.java:
```diff
@@ -24,7 +24,7 @@ public ResponseEntity<ErrorResponse> handleBadRequestException(BadRequestExcepti
     }
 
-    @ExceptionHandler(value = {InternalServerException.class})
+    @ExceptionHandler(value = {InternalServerException.class, RuntimeException.class})
```
RuntimeException에 대한 핸들러는 구현이 안되어있어서 내부적으로 저희가 정의한 스펙에 벗어난 예외가 발생할 경우 핸들러를 호출하지 않고 에러가 발생할 거 같다는 생각이 들더군요. 그래서 우선은 Internal Server Error 스펙과 동일하다고 생각하고 핸들러가 다루는 예외 클래스로 RuntimeException을 추가하였습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -0,0 +1,63 @@
+        this.objectMapper = objectMapper;
+    }
+
+    @PostMapping("/signup")
```
가게를 생성하는데, 해당 URL 뜻이 적절하지 않은 듯 싶습니다!

y00njinuk replied:
흠... 어떤게 적절할까요? create..?

rnjstmdals6 replied:
나쁘지 않은 것 같습니다 ㅎㅎ

y00njinuk replied:
🆗

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @PostMapping("/signup")
+    public ResponseEntity<StoreResponseDTO> signup(@RequestBody StoreSignupRequestDTO dto) {
```
유효성 체크를 하는 어노테이션을 추가해주세요!

rnjstmdals6 replied:
가게를 생성, 수정, 삭제하는 건 로그인된 사장님만 가능하도록 해야하지 않을까 싶습니다.

y00njinuk replied:
> 유효성 체크를 하는 어노테이션을 추가해주세요!
- @Valid 어노테이션을 추가하라는 말씀이신가요?

> 가게를 생성, 수정, 삭제하는 건 로그인된 사장님만 가능하도록 해야하지 않을까 싶습니다.
- 이 부분은 권한이랑 관련이 있는거 같네요. 조금 조사가 필요할 거 같은데.. 관련 키워드를 알려주실 수 있을까요?

rnjstmdals6 replied:
1. 넵 맞습니다.
2. @AuthenticationPrincipal 관련해서 찾아보시는 걸 추천합니다!

y00njinuk replied:
넵 말씀해주시는 걸로 적용하도록 하겠습니다.

y00njinuk replied:
create로 우선 수정하였습니다. 
AuthenticationPrincipal 어노테이션을 좀 더 찾아보고 커밋을 올리도록 하겠습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -0,0 +1,63 @@
+                .body(res);
+    }
+
+    @GetMapping("/find")
```
@GetMapping("/find/{id}")
    public ResponseEntity<StoreResponseDTO> findById(@RequestParam Long id) {
        StoreResponseDTO res = storeService.findById(id));

        return ResponseEntity
                .status(HttpStatus.OK)
                .body(res);
    }
    
    이렇게 작성하는 건 어떠신가요?

y00njinuk replied:
URL path에 요청 정보를 넣느냐 parameter에 정보를 넣느냐의 차이인 거 같은데 추후에 여러 개의 정보로 조회 요청이 들어올 거 같아서 parameter로 정의하였습니다. path를 활용하면 이와 같은 경우에는 적절치 않은 거 같아서 말씀해주신 방법대로 구현하지 않았습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -0,0 +1,63 @@
+                .body(res);
+    }
+
+    @PutMapping("/update")
```
일부 변경이라면 @Patch 매핑이 의미가 더 맞을 것 같습니다!

y00njinuk replied:
넵 수정하도록 하겠습니다.

y00njinuk committed:
fix: fix error

[issue-key] DS-56
- 테스트 실패 조치
- 발생한 에러는 다음과 같으며 정확한 원인 분석이 필요
- java.lang.IllegalStateException: Could not resolve parameter [0] in public org.springframework.http.ResponseEntity<com.dangol.dangolsonnimbackend.errors.RestApiExceptionHandler$ErrorResponse>

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/boss/dto/BossSigninReqeustDTO.java:
```diff
@@ -10,6 +11,7 @@
 
 @Getter
 @Setter
+@NoArgsConstructor
```
음.. 이게 테스트 에러가 나서 로그를 보니 생성자를 추가해달라는 내용이 있네요.
정확한 원인은 분석이 필요해보입니다.

rnjstmdals6 replied:
Request Body로 들어오는 JSON을 해당 DTO 객체로 역직렬화할 때 기본 생성자를 사용하기 때문입니다.

y00njinuk replied:
😮컥 ... 그전엔 왜 오류가 발생하지 않은거죠!? 전 main 브랜치에서 pull 한게 전부인데..

rnjstmdals6 replied:
서비스 테스트할 때 말씀이신가요!?

y00njinuk replied:
- BossControllerTest > givenInvalidCredentials_whenAuthenticate_thenFail() FAILED
- BossControllerTest > givenValidCredentials_whenAuthenticate_thenSucceed() FAILED
- BossControllerTest > givenBossEmail_whenUpdate_thenReturnBoss() FAILED

위 테스트가 이전엔 왜 실패하지 않은거죠?

rnjstmdals6 replied:
테스트 에러가 나서 @NoArgsConstructor 를 추가하니 해결된 부분 아닌가요?

y00njinuk replied:
네 해결은 됐는데 왜 발생했는지 알 수가 없어서요...

y00njinuk committed:
refactor: update url path

[issue-key] DS-56
- 가게 생성을 요청하는 URL 경로 변경

y00njinuk committed:
feat: validate StoreSignupRequestDTO for request

[issue-key] DS-56
- 가게 생성을 요청하는 DTO에 대하여 validate 추가

y00njinuk committed:
refactor: modify http method when request for update store

[issue-key] DS-56
- 가게 수정을 요청하는 HTTP 메서드 수정

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -29,7 +30,7 @@ public StoreController(StoreService storeService, ObjectMapper objectMapper) {
 
     @PostMapping("/create")
-    public ResponseEntity<StoreResponseDTO> signup(@RequestBody StoreSignupRequestDTO dto) {
+    public ResponseEntity<StoreResponseDTO> signup(@Valid @RequestBody StoreSignupRequestDTO dto) {
```
Valid 어노테이션을 추가하였습니다.

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/controller/StoreController.java:
```diff
@@ -47,7 +47,7 @@ public ResponseEntity<StoreResponseDTO> findById(@RequestParam Map<String, Strin
     }
 
-    @PutMapping("/update")
+    @PatchMapping("/update")
```
HTTP 메서드를 Patch로 수정하였습니다.
